pax/plugins/signal_processing/DesaturatePulses.py

```diff
@@ -27,71 +27,117 @@ def transform_event(self, event):
-                continue
-
-            # Reconstruct the waveform in the saturated region according to this ratio.
-            # The waveform should never be reduced due to this (then the correction is making things worse)
-            w[saturated] = np.clip(sumw[saturated] * ratio, w[saturated], float('inf'))
-
-            # Convert back to raw ADC counts and store the corrected waveform
-            # Note this changes the type of pulse.w from int16 to float64: we don't have a choice,
-            # int16 probably can't contain the large amplitudes we may be putting in.
-            # As long as the raw data isn't saved again after applying this correction, this should be no problem
-            # (as in later code converting to floats is anyway the first step).
-            w /= adc_to_pe(self.config, pulse.channel)
-            w = self.reference_baseline - w - pulse.baseline
-
-            pulse.raw_data = w
+            for peak_i, _where_saturated in enumerate(_where_saturated_list):
+                try:
+                    first_saturated = _where_saturated.min()
+                    last_saturated = _where_saturated.max()
+                except (ValueError, RuntimeError, TypeError, NameError):
+                    continue
+
+                # Select a reference region just before the start of the saturated region
+                reference_slice = slice(max(0, first_saturated - self.config['reference_region_samples']),
+                                        first_saturated)
+
+                # Compute the (gain-weighted) sum waveform of the non-saturated pulses
+                min_left = min([p.left for p in other_pulses + [pulse]])
+                max_right = max([p.right for p in other_pulses + [pulse]])
+                sumw = np.zeros(max_right - min_left + 1)
+                for p in other_pulses:
+                    offset = p.left - min_left
+                    sumw[offset:offset + len(p.raw_data)] += self.waveform_in_pe(p)
+
+                # Crop it to include just the part that overlaps with this pulse
+                offset = pulse.left - min_left
+                sumw = sumw[offset:offset + len(pulse.raw_data)]
+
+                # Compute the ratio of this channel's waveform / the nonsaturated waveform in the reference region
+                w = self.waveform_in_pe(pulse)
+                if len(sumw[reference_slice][sumw[reference_slice] > 1]) \
+                        < self.config['reference_region_samples_treshold']:
+                    # the pulse is saturated, but there are not enough reference samples to get a good ratio
+                    # This actually distinguished between S1 and S2 and will only correct S2 signals
+                    continue
+
+                ratio = w[reference_slice].sum()/sumw[reference_slice].sum()
+
+                # not < is preferred over >, since it will catch nan
+                if not ratio < self.config.get('min_reference_area_ratio', 1):
```

Comment:
would be good to add a link to a plot for each exception, and put them in your note.